include/zephyr/sys/fdtable.h

```diff
@@ -21,8 +35,14 @@ extern "C" {
  * Currently all operations beyond read/write/close go thru ioctl method.
  */
 struct fd_op_vtable {
-	ssize_t (*read)(void *obj, void *buf, size_t sz);
-	ssize_t (*write)(void *obj, const void *buf, size_t sz);
+	union {
+		ssize_t (*read)(void *obj, void *buf, size_t sz);
```

Comment:
I have a little concern about the unions here resulting in some difficult to find bugs for future implementers. I guess the other option would be to have these separate, but that would take additional space, but is that a lot?

Reply:
Pushing an extra, unused, argument onto the stack and popping it back off shouldn't be too much of a concern I would think. But up to 16 extra (unneeded / unused) bytes per file descriptor might be a lot of wasted space, especially given that most seekable file types should eventually adopt the `_offset()` variants, and it's they are otherwise identical.

It should be noted as well, that offsets do not work apply to sockets, pipes, etc, just regular files and memory.

Reply:
I guess I was more concerned with the subtle semantic difference. If the implementation is an offset, but the caller is expecting without, that offset argument will be read as garbage.

Reply:
Interestingly, `lseek()` must be able to set the file descriptor offset to an arbitrary value. Kind of mind boggling.

> The lseek() function shall allow the file offset to be set beyond the end of the existing data in the file.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/lseek.html

Consequently, read and write implementations must be able to decide if an offset is valid or invalid and react accordingly.

So they should be able to handle garbage.